### PR TITLE
Handle `cluster.max_shards_per_node` in YAML config

### PR DESCRIPTION
**Pull request by AthenaEryma**
Prior to this commit, `cluster.max_shards_per_node` is not correctly handled
when it is set via the YAML config file, only when it is set via the Cluster
Settings API.

This commit refactors how the limit is implemented, both to enable correctly
handling the setting in the YAML and to more effectively centralize the logic
used to enforce the limit. The logic used to apply the limit, as well as the
setting value, has been moved to the new `ShardLimitValidator`.

Relates https://github.com/elastic/elasticsearch/issues/40803

### CODE REVIEW

**Comment by gwbrown** [comment]
I'm thinking `ShardLimitChecker` or `ShardLimitValidator` might be better names than `ShardLimitService`. If anyone has suggestions on that front, I'm happy to hear them.

File: server/src/main/java/org/elasticsearch/indices/ShardLimitService.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.Index;
+
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.elasticsearch.cluster.metadata.IndexMetadata.INDEX_NUMBER_OF_SHARDS_SETTING;
+
+public class ShardLimitService {
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/indices/ShardLimitService.java:36]: Can you add some javadocs please?

File: server/src/main/java/org/elasticsearch/indices/ShardLimitService.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    private static int getTotalShardCount(ClusterState state, Index index) {
+        IndexMetadata indexMetadata = state.metadata().index(index);
+        return indexMetadata.getNumberOfShards() * (1 + indexMetadata.getNumberOfReplicas());
+    }
+
+    /**
+     * Checks to see if an operation can be performed without taking the cluster over the cluster-wide shard limit. Adds a deprecation
+     * warning or returns an error message as appropriate
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/indices/ShardLimitService.java:104]: I don't think any deprecation warning is added?

File: server/src/main/java/org/elasticsearch/indices/ShardLimitService.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+    /**
+     * Validates whether a list of indices can be opened without going over the cluster shard limit.  Only counts indices which are
+     * currently closed and will be opened, ignores indices which are already open.
+     *
+     * @param currentState The current cluster state.
+     * @param indices The indices which are to be opened.
+     * @throws ValidationException If this operation would take the cluster over the limit and enforcement is enabled.
+     */
+    public void validateShardLimit(ClusterState currentState, Index[] indices) {
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/indices/ShardLimitService.java:83]: Super minor, but I was confused until I read the docs, maybe rename the parameter "indicesToBeOpened"?

**Review by dakrone — APPROVED** [review]
LGTM, I left some minor comments, I like `ShardLimitValidator` much more than calling it a "service", so that's my pick.

File: server/src/main/java/org/elasticsearch/indices/ShardLimitService.java (hunk 2 above)
Comment (gwbrown) [reply server/src/main/java/org/elasticsearch/indices/ShardLimitService.java:104]: This comment is referencing a deprecation warning that's present in the 7.x branch, not master. I'll remove the reference to the deprecation warning in master but we'll need to keep emitting the deprecation warning in the backport (which I believe there are tests for).

File: server/src/main/java/org/elasticsearch/indices/ShardLimitService.java (hunk 3 above)
Comment (gwbrown) [reply server/src/main/java/org/elasticsearch/indices/ShardLimitService.java:83]: 👍 This was moved from another class which had clearer context for this, I'll update the name.

**Pull request merged**